### PR TITLE
usb: add Chapter 9 (USB Device Framework) header

### PR DESCRIPTION
**Pull request by jfischer-no**
Add header file where all defines and structures
from Chapter 9 (USB Device Framework) should be included.

Replace all macros and types in the tree with the new ones from
usb/usb_ch9.h header.

Deprecate usb_common.h and usbstruct.h headers.
Deprecate DESCRIPTOR_TYPE_BOS macro.

### CODE REVIEW

**Review by carlescufi — APPROVED** [review]

File: include/usb/usbstruct.h
Comment (mbolivar-nordic) [inline include/usb/usbstruct.h:35]: Shouldn't this be inside of the include guard?

**Comment by cfriedt** [comment]
@jfischer-no - could you fix your commit message? Is there anything else preventing this from being merged?

**Comment by mbolivar-nordic** [comment]
>  Is there anything else preventing this from being merged?

I think @jfischer-no should at least respond to the portability concerns I outlined.

File: include/usb/usb_ch9.h
Comment (jfischer-no) [reply include/usb/usb_ch9.h:115]: I however find it easier to read, even if one need it inverted, `return foo == bar ? false : true;`

> Is there some MISRA thing where that requires a cast or something?

I do not know about MISRA. The type of 'foo == bar' is int, value 1 or 0, just as bool (_Bool) can only have these values.

File: include/usb/usb_ch9.h
Comment (carlescufi) [reply include/usb/usb_ch9.h:35]: @mbolivar-nordic while you are entirely right in principle, multiple protocol stacks (including Bluetooth) use this style of bitfield declaration, and so does Linux and other close-to-metal open source projects. In practice all compilers respect this idiom, and so I would be in favor of keeping it here as well.

File: include/usb/usbstruct.h
Comment (mbolivar-nordic) [reply include/usb/usbstruct.h:35]: > Shouldn't this be inside of the include guard?

Ping on this one

File: include/usb/usb_ch9.h
Comment (mbolivar-nordic) [reply include/usb/usb_ch9.h:115]: I think it is verbose and harder to read, but OK, you're the maintainer

File: include/usb/usb_ch9.h
Comment (mbolivar-nordic) [reply include/usb/usb_ch9.h:35]: > In practice all compilers respect this idiom

If that is true, then okay.

> Given that using bitfields gives the compiler the opportunity to optimize, I would be in favor of keeping them.

Here I think a citation is needed that this actually results in better optimized code.

**Review by mbolivar-nordic — APPROVED** [review]

File: include/usb/usb_ch9.h
Comment (jfischer-no) [reply include/usb/usb_ch9.h:115]: I will change it.

File: include/usb/usbstruct.h
Comment (jfischer-no) [reply include/usb/usbstruct.h:35]: I want to test something, I will remove or justify it at the next push.

File: include/usb/usb_ch9.h
Comment (emob-nordic) [inline include/usb/usb_ch9.h:2]: Is it still valid?

File: include/usb/usb_ch9.h
Comment (emob-nordic) [inline include/usb/usb_ch9.h:12]: You could mention those exceptions here.

**Review by eobalski — APPROVED** [review]

**Comment by cfriedt** [comment]
@jfischer-no - please mark any comments resolved if they are resolved, fix the merge conflict, and any compliance issues. I think everyone wants this to be in, so should be ready to merge when all of that is done.

File: include/usb/usb_ch9.h
Comment (jfischer-no) [reply include/usb/usb_ch9.h:2]: Yes, this commit is from #27045

File: include/usb/usb_ch9.h
Comment (jfischer-no) [reply include/usb/usb_ch9.h:12]: I will improve the documentation in subsequent PR's.

File: include/usb/usbstruct.h
Comment (jfischer-no) [reply include/usb/usbstruct.h]: Moved inside include guard

**Pull request merged**
